pkg/query-service/app/cloudintegrations/controller.go

```diff
@@ -433,37 +429,43 @@ func (c *Controller) GetServiceDetails(
 		}
 
 		if config != nil {
-			service.Config = config
+			details.Config = config
 
 			enabled := false
 			if config.Metrics != nil && config.Metrics.Enabled {
 				enabled = true
 			}
 
 			// add links to service dashboards, making them clickable.
-			for i, d := range service.Assets.Dashboards {
+			for i, d := range definition.Assets.Dashboards {
 				dashboardUuid := c.dashboardUuid(
 					cloudProvider, serviceId, d.Id,
 				)
 				if enabled {
-					service.Assets.Dashboards[i].Url = fmt.Sprintf(
-						"/dashboard/%s", dashboardUuid,
-					)
-				} else {
-					service.Assets.Dashboards[i].Url = ""
+					definition.Assets.Dashboards[i].Url = fmt.Sprintf("/dashboard/%s", dashboardUuid)
 				}
 			}
 		}
 	}
 
-	return service, nil
+	return &details, nil
 }
 
 type UpdateServiceConfigRequest struct {
 	CloudAccountId string                   `json:"cloud_account_id"`
 	Config         types.CloudServiceConfig `json:"config"`
 }
 
+func (u *UpdateServiceConfigRequest) Validate(def *services.Definition) error {
+	if def.Id != services.S3Sync && u.Config.Logs != nil && u.Config.Logs.S3Buckets != nil {
+		return fmt.Errorf("s3 buckets can only be added to service-type[%s]", services.S3Sync)
```

Comment:
use the new errors package, check new packages inside modules to get idea about how to use them.